src/mscorlib/shared/System/HashCode.cs

```diff
@@ -359,75 +356,78 @@ private void Add(int value)
                 _queue2 = val;
             else if (position == 2)
                 _queue3 = val;
-            else // == 3
+            else // position == 3
             {
-                // length smaller than 4?
-                if ((_length ^ position) == 0)
+                if (previousLength == 3)
                     Initialize(out _v1, out _v2, out _v3, out _v4);
 
                 _v1 = Round(_v1, _queue1);
                 _v2 = Round(_v2, _queue2);
                 _v3 = Round(_v3, _queue3);
                 _v4 = Round(_v4, val);
             }
-
-            // Throw for more than uint.MaxValue fields.
-            _length = checked(_length + 1);
         }
 
         public int ToHashCode()
         {
-            uint position = _length % 4;
+            uint length = _length;
+            uint position = length % 4;
 
-            // If the length is less than 3, _v1 -> _v4 don't contain
-            // anything yet. xxHash32 treats this differently.
+            // If the length is less than 4, _v1 to _v4 don't contain anything
+            // yet. xxHash32 treats this differently.
 
-            uint hash = (_length ^ position) == 0
-                ? MixEmptyState()
-                : MixState(_v1, _v2, _v3, _v4);
+            uint hash;
+            if (length < 4)
+                hash = MixEmptyState();
+            else
+                hash = MixState(_v1, _v2, _v3, _v4);
 
-            // Multiply by 4 because we've been counting in ints, not
-            // bytes.
+            // Multiply by 4 because we've been counting in bytes, not ints.
 
-            hash += _length * 4;
+            hash += length * 4;
 
             // Mix what remains in the queue
-            // Switch can't be inlined right now, so use as few
-            // branches as possible instead.
 
-            if (position > 0)
-            {
-                hash = QueueRound(hash, _queue1);
-                if (position > 1)
-                {
-                    hash = QueueRound(hash, _queue2);
-                    if (position > 2)
-                        hash = QueueRound(hash, _queue3);
-                }
-            }
+            // Switch can't be inlined right now, so emulate case statement
+            // fallthrough using goto.
+
+            // position refers to the *next* queue position in this method, so
+            // position == 1 means that _queue1 is populated; _queue2 would have
+            // been populated on the next call to Add.
 
+            if (position == 0) goto mixFinal;
```

Comment:
This has the same effect as the nested-ifs you had earlier, actually. I had something different in mind for `goto`, but now I don't think it will let us get the same perf as if we could use `switch`. You can revert this part to how it was before.

Have you tried marking `ToHashCode` with AggressiveInlining and using `switch`? Does `switch` generate more code than ifs?

Reply:
According to my tests, the JIT will *never* inline a method with `switch`, regardless of `AggressiveInlining`.

Reply:
Yes, this is a limitation in the current inliner.